### ISSUE TITLE
Modification check failed for material for pipelines that polling is disabled and or are paused .

### ISSUE THREAD

**Issue by childsc** [gocd/gocd#6030]
##### Issue Type
- Bug Report

##### Summary

We in our team manage about 800 pipelines for different teams within the company I work for.

Unfortunately, we are receiving a lot of warning and errors with the "message" starting "Modification check failed for material ..."

All these pipelines are for teams where the branch which they were using is no longer available and they have not configured the new branch in Go (possibly because there is no new branch to work from yet)

I was hoping that if you were to either (or both) Pause a pipeline and/or NOT select polling in the pipeline configuration it would stop this behavior.

The teams do not want to keep deleting the materials and adding dummy materials to pipelines and then needing to recreate the materials when needed as this would mean additional extra work.

###### Basic environment details

* Go Version: GoCD Version: 18.12.0 (8222-e6778f1cc84bf91e323d337c876046167da36985).
* JAVA Version:  OpenJDK 11.0.1
* OS: Linux 3.13.0-147-generic
* Browser vendor and version (if relevant): NA

###### Additional Environment Details
Example of error:

message : Modification check failed for material: URL: http://USERNAME:******@[URL PATH TO REPO] Branch: testing
detail  : Failed to run git clone command STDERR: Cloning into '/var/lib/go-server/pipelines/flyweight/6b10e850-5b6a-4bd2-acc5-9972543d87ad'...
          STDERR: fatal: Authentication failed for 'http://USERNAME:******@[URL PATH TO REPO]'
level   : ERROR
time    : 3/27/2019 10:45:25 AM

message : Modification check failed for material: URL: http://USERNAME:******@[URL PATH TO REPO], Branch: development
detail  : Working directory: pipelines/flyweight/10667df4-940b-4fdd-b534-2a80e07028d4
          STDERR: fatal: Authentication failed for 'http://USERNAME:******@[URL PATH TO REPO]'
level   : ERROR
time    : 3/27/2019 10:44:36 AM

##### Steps to Reproduce

1.  Create Test Branch for repo
2.  Configure Go to use above created branch as a material
3.  Disable polling for the material 
4.  Remove Branch from repo
5.  Receive warning message

##### Expected Results
Disable polling of material should mean repo is not verified.

##### Actual Results
We get a "Modification check failed for material:" warning 

##### Possible Fix
When disabling polling materials are not actually checked for existence. I would hope this should also boost the server performance?

##### Log snippets
message : Modification check failed for material: URL: http://USERNAME:******@[URL PATH TO REPO] Branch: testing
detail  : Failed to run git clone command STDERR: Cloning into '/var/lib/go-server/pipelines/flyweight/6b10e850-5b6a-4bd2-acc5-9972543d87ad'...
          STDERR: fatal: Authentication failed for 'http://USERNAME:******@[URL PATH TO REPO]'
level   : ERROR
time    : 3/27/2019 10:45:25 AM

message : Modification check failed for material: URL: http://USERNAME:******@[URL PATH TO REPO], Branch: development
detail  : Working directory: pipelines/flyweight/10667df4-940b-4fdd-b534-2a80e07028d4
          STDERR: fatal: Authentication failed for 'http://USERNAME:******@[URL PATH TO REPO]'
level   : ERROR
time    : 3/27/2019 10:44:36 AM

**Comment by arvindsv** [comment]
/cc @ankitsri11 Can you take a look at this?

I agree that setting `autoUpdate` to `false` (disabling polling) should mean that the material is not polled.

**Comment by childsc** [comment]
Thanks @arvindsv and @ankitsri11  I look forward to seeing an update on this. :)

**Comment by ankitsri11** [comment]
@childsc I tried to replicate the steps you shared but couldn't able to reproduce the warning locally. I tried on both GoCD 18.12.0 and 19.2.0.

Here are the steps which I followed:

1. Create Test Branch for repo
2. Created a pipeline with the same repo material pointing to Test branch and disabled poll for new changes.

```
 <pipeline name="Sample">
      <materials>
        <git url="https://github.com/ankitsri11/some-repo.git" branch="test" autoUpdate="false" />
      </materials>
      <stage name="defaultStage">
        <jobs>
          <job name="defaultJob">
            <tasks>
              <exec command="ls" />
            </tasks>
          </job>
        </jobs>
      </stage>
    </pipeline>
```
3. Unpaused the pipeline and pipeline din't trigger as poll for new changes were disabled.
4. Remove the Test branch from repo 

However, when I tried below I got the warning message and its an expected behaviour as I deleted the branch first and then disabled poll for new changes.

1. Created pipeline with the test branch material
2. Deleted the branch
3. Disable polling for the material
4. Receive warning message

```
2019-03-29 12:02:49,669 WARN  [115@MessageListener for MaterialUpdateListener] MaterialDatabaseUpdater:112 - [Material Update] Modification check failed for material: URL: https://github.com/ankitsri11/some-repo.git, Branch: test
java.lang.RuntimeException: Failed to run git clone command STDERR: Cloning into '/Users/ankitsri/Library/Application Support/Go Server/go-server-18.12.0/pipelines/flyweight/8f078af6-6b05-496b-805c-9872b44cf9d5'...
STDERR: fatal: Remote branch test not found in upstream origin
```

**Comment by childsc** [comment]
Hello.
Thank you @ankitsri11 can you tell me if this is expected? or are you considering this still an issue? 

thanks for your time.

**Comment by arvindsv** [comment]
@childsc 

Here's what would be expected:

1. Delete a branch.
2. See a message in the logs / server health message.
3. Disable polling of that material as shown above.
4. The message goes away and does not come back, since polling is disabled.

Here's what would be unexpected and probably an issue:

1. You disable polling on a material.
2. Delete a branch or do anything on it.
3. _There's a message shown_ [will probably be a bug]

That message should show only when polling is enabled OR someone force triggers that pipeline, causing that material to be checked.


In your original post (above), you'd said:

> I was hoping that if you were to either (or both) Pause a pipeline and/or NOT select polling in the pipeline configuration it would stop this behavior.

What Ankit is saying is that he thinks that the message should not be showing if `autoUpdate` is set to false, for that material. Are you saying you're seeing it even if `autoUpdate` is false? If so, can you provide a small example or a small video showing how it can reproduced?

Thank you!

**Comment by childsc** [comment]
Yes we are still seeing this problem. I have attached short documentation with screen shots involving 2 tests. 

Summary of tests:
1. With Branch Test_Materials:
    Create GoCD config with materials
    Keep polling enabled
    Remove repo
    Disable polling

2. With Branch Test-Materials:
    Create GoCD config with materials
    Disable Polling
    Remove repo
    Enable polling
    Disable Polling

I would expect that when polling is disabled then within a minute or two the error should disappear from the console or via the results from the api.  Instead we see some errors dated from months ago.

I have noticed that it can be hard to track an error to a particular pipeline without searching for the material in the config.xml.  The only way we can do it in a nice report is to get the Errors from the API and then Get the config.xml with the api and just search each pipeline for the material :)
[GoCDSupportingEvidence.docx](https://github.com/gocd/gocd/files/3043160/GoCDSupportingEvidence.docx)

Kind Regards

**Comment by ankitsri11** [comment]
@childsc thank you for sharing the detailed steps. It was useful in troubleshooting the issue. I was able to replicate the issue mentioned by you. In a case when the branch is deleted and poll for new changes is disabled afterward, the warning message from the GoCD UI and API support page doesn't go away. However, GoCD stops logging that error in the logs. 

I'll have someone from the team to look at it and see if that can be cleared.

A workaround, for now, would be to restart your GoCD server. I restarted my instance after changes (changes to enable warning message; delete branch and disable poll for new changes) and the warning messages went away.
